### PR TITLE
FIREFLY-1304: Implement spectral redshift correction

### PR DESCRIPTION
**Pull request by jaladh-singhal**
Implements [FIREFLY-1304](https://jira.ipac.caltech.edu/browse/FIREFLY-1304)

On a high-level:
- Made spectral frame data in the table available to the chart data as `fireflyData`
- Added UI for spectral frame and redshift input in Spectrum options
  -  Also updated it based on early feedback from scientists
- Made changes in spectral frame UI inputs trigger changes in X ColumnOrExpression and X Label - aka redshift correction
  - Also apply those changes to other traces of the chart
- [x] Combine expressions for unit conversion and redshift correction, also fix label combination
- [x] Make it work for CUSTOM or other ref pos too
- [x] Chart should be initialized based on initial state of spectral frame options

## Testing
https://fireflydev.ipac.caltech.edu/firefly-1304-redshift-correction/firefly/

Created test files for each of the following 3 case available at https://github.com/jaladh-singhal/z-correction-data/

1. TOPOCENTER - Download [Spitzer_IRS_SL_PBCD-topocenter.vot](https://github.com/jaladh-singhal/z-correction-data/blob/main/Spitzer_IRS_SL_PBCD-topocenter.vot) and upload it to firefly. Open chart options, change spectral frame to rest frame and then in resulting options, choose user-specifed one. Enter 0.5, and apply, check that x axis data and label changed in the chart for all traces.

   Also note that the rest frame options have values that are present in table metadata (retrievable via clicking info button in table). And changing value in dropdown (rest/observed) preserves selection in rest frame radio group options as well as input field.

   Stress test it for other edge cases that can occur due to selection in other fields in spectrum options.

2. CUSTOM - Download [Spitzer_IRS_SL_PBCD-custom.vot](https://github.com/jaladh-singhal/z-correction-data/blob/main/Spitzer_IRS_SL_PBCD-custom.vot) and upload it to firefly. Open chart options, spectral frame should have a custom redshift option selected by default. Change options and play around like in TOPOCENTER case. Also note that formula has multiplication factor in this case.

3. Other - Download [Spitzer_IRS_SL_PBCD-barycenter.vot](https://github.com/jaladh-singhal/z-correction-data/blob/main/Spitzer_IRS_SL_PBCD-barycenter.vot) and upload it to firefly. Open chart options, spectral frame shouldn't be editable as per design and x-axis label in the chart should show the spectral frame.

### CODE REVIEW

**Comment by robyww** [comment]
#### Testing

- upload the fits file below, choose first image.
- Use the cube extract: image toolbar -> tools dropdown -> Extract: drill.
- click on point
- dialog come up: click on Pin Chart/Table.
- the modify the pin spectrum by choosing rest frame.
- I get the error in the screen shot


Screenshot 2023-11-02 at 2 40 28 PM


unzip and upload this fits file:
[ cube.fits.zip](https://github.com/Caltech-IPAC/firefly/files/13244270/cube.fits.zip)

**Review by robyww — COMMENTED** [review]
Code looks good see my testing comment.

**Comment by jaladh-singhal** [comment]
> Code looks good see my testing comment.

Thanks. I just tested it, it's working fine for me. Can you hover over error icon and tell me what it is?

File: src/firefly/js/charts/dataTypes/SpectrumUnitConversion.js [hunk 1]
```diff
@@ -176,3 +176,16 @@ const UnitXref = {
...
+/**
+ * returns X axis label combining spectral frame and redshift information
+ * @param {string} cname         the name of the column being evaluated
+ * @param {string} unit          the unit to get the info for
+ * @param {string} sfLabel       the label of spectral frame selected
+ * @param {string} redshiftLabel the label of redshift if rest frame, optional
+ * @returns {string}
+ */
+export const getCombinedXLabel = (cname, unit, sfLabel, redshiftLabel='') => {
+    const unitLabel = getUnitInfo(unit, cname).label;
```
Comment (loitly) [inline src/firefly/js/charts/dataTypes/SpectrumUnitConversion.js:189]: I understand why you named it such.  But, it makes more sense to just call it getXLabel, or getDefaultXLabel.  In the function, you can comment on what was taken into consideration when creating the label, which you already have.  There may be more in the future.

**Review by loitly — APPROVED** [review]
Good work!  Clean code and it works great.  I noticed the validation error Trey mentioned once.  But, it looks like a rendering issue, because I clicked apply, and it did not complain.  `Combine Chart` should be a separate ticket.  Science team need to describe what should happen when combining.  Also, the unit conversion bug Justin mentioned is an existing bug.

File: src/firefly/js/charts/dataTypes/SpectrumUnitConversion.js (hunk 1 above)
Comment (jaladh-singhal) [reply src/firefly/js/charts/dataTypes/SpectrumUnitConversion.js]: Agreed, done!

**Comment by jaladh-singhal** [comment]
> I noticed the validation error Trey mentioned once. But, it looks like a rendering issue, because I clicked apply, and it did not complain.

I traced the call stack and figured that it was happening due to stale validation state. Had to trigger revalidation explicitly because we are changing state directly (without UI). It's fixed now @robyww 

> Combine Chart should be a separate ticket. Science team need to describe what should happen when combining.

Yes, that's what I was thinking too and didn't do anything about combine chart.

> Also, the unit conversion bug Justin mentioned is an existing bug.

Ok, good to know.
